src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -198,47 +198,48 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
                 hasErrors = true;
             }
 
-            BoundIndexerAccess? indexerAccess = null;
-            MethodSymbol? sliceMethod = null;
+            BoundExpression? indexerAccess = null;
             BoundPattern? pattern = null;
+            BoundSlicePatternReceiverPlaceholder? receiverPlaceholder = null;
+            BoundSlicePatternRangePlaceholder? argumentPlaceholder = null;
 
             // We don't require the type to be sliceable if there's no subpattern.
             if (node.Pattern is not null)
             {
+                receiverPlaceholder = new BoundSlicePatternReceiverPlaceholder(node, GetValEscape(inputType, inputValEscape), inputType) { WasCompilerGenerated = true };
+                var systemRangeType = GetWellKnownType(WellKnownType.System_Range, diagnostics, node);
+                argumentPlaceholder = new BoundSlicePatternRangePlaceholder(node, systemRangeType) { WasCompilerGenerated = true };
+
                 TypeSymbol sliceType;
                 if (inputType.IsErrorType())
                 {
                     hasErrors = true;
                     sliceType = inputType;
                 }
-                else if (inputType.IsSZArray())
-                {
-                    sliceType = inputType;
-                }
-                else if (TryPerformPatternIndexerLookup(node, inputType, argIsIndex: false, out indexerAccess, out Symbol? patternSymbol, lengthProperty: out _, diagnostics))
+                else
                 {
-                    if (patternSymbol is MethodSymbol method)
-                    {
-                        sliceMethod = method;
-                        sliceType = method.ReturnType;
-                    }
-                    else
+                    var analyzedArguments = AnalyzedArguments.GetInstance();
+                    analyzedArguments.Arguments.Add(argumentPlaceholder);
+
+                    indexerAccess = BindElementAccessCore(node, receiverPlaceholder, analyzedArguments, diagnostics).MakeCompilerGenerated();
+                    indexerAccess = CheckValue(indexerAccess, BindValueKind.RValue, diagnostics);
+                    Debug.Assert(indexerAccess is BoundIndexerAccess or BoundIndexOrRangePatternIndexerAccess or BoundArrayAccess or BoundBadExpression or BoundDynamicIndexerAccess);
+                    analyzedArguments.Free();
+
+                    if (!systemRangeType.HasUseSiteError)
                     {
-                        Debug.Assert(indexerAccess is not null);
-                        sliceType = indexerAccess.Type;
+                        _ = GetWellKnownTypeMember(WellKnownMember.System_Range__ctor, diagnostics, syntax: node);
```

Comment:
>_ = GetWellKnownTypeMember(WellKnownMember.System_Range__ctor, diagnostics, syntax: node); [](http://example.com/codeflow?start=24&length=90)

I guess it is fine to check this, however it looks like lowering handles error reporting for this.